src/Algebra/Graph/AdjacencyMap/Algorithm.hs

```diff
@@ -213,32 +240,87 @@ dfs vs = concatMap flatten . dfsForestFrom vs
 -- 'isSubgraphOf' ('vertices' $ reachable x y) y == True
 -- @
 reachable :: Ord a => a -> AdjacencyMap a -> [a]
-reachable x = concat . bfs [x]
+reachable x = dfs [x]
+
+type Cycle = NonEmpty 
+data Entry a = Entered a | Exited a
+type TopSort a = Either (Cycle a) [a]
+type ParentTable a = Map.Map a (Entry a)
+data S a = S { table :: !(ParentTable a), order :: [a] }
+
+topSort' :: (Ord a, MonadState (S a) m, MonadCont m)
+         => AdjacencyMap a -> m (TopSort a)
+topSort' g = callCC $ \cyclic ->
+  do let vertices = map fst $ Map.toDescList $ adjacencyMap g
+         adjacent = Set.toDescList . flip postSet g
+         dfs prev curr =
+           do enter prev curr
+              forM_ (adjacent curr) $ \next ->
+                nodeState next >>= \case
+                  Nothing -> dfs curr next
+                  Just (Exited _) -> return ()
+                  _ -> cyclic . Left . retrace curr next =<< gets table
+              exit curr
+     forM_ vertices $ \v -> nodeState v >>= \case
+       Nothing -> dfs undefined v
```

Comment:
Ouch. Is it possible to avoid `undefined` here?

Reply:
I added two constructors to the `Entry` type to avoid this. No undefined, but more lines of code in the definitions

Reply:
Hopefully it doesn't affect performance too much

Reply:
The CI result is still `dff: 0.50 (good)`, so performance seems to be unaffected!

What about your own experiments, do you see any degradation of performance because of the addition of new constructors?

Reply:
There appears to be no impact!

Reply:
Awesome!